cachi2/core/package_managers/bundler/parser.py

```diff
@@ -194,7 +222,23 @@ def parse_lockfile(package_dir: RootedPath) -> ParseResult:
     result: ParseResult = []
     for dep in dependencies:
         if dep["type"] == "rubygems":
-            result.append(GemDependency(**dep))
+            if f'{dep["name"]}-{dep["version"]}' != dep["full_name"]:
+                log.warning("Found a binary dependency %s", dep["full_name"])
+                if allow_binary:
+                    log.warning(
+                        "Downloading binary dependency %s because 'allow_binary' is set to True",
+                        dep["full_name"],
+                    )
+                    result.append(GemPlatformSpecificDependency(**dep))
+                else:
+                    # No need to force a platform if we skip the packages.
+                    log.warning(
+                        "Skipping binary dependency %s because 'allow_binary' is set to False."
+                        " This will likely result in an unbuildable package.",
```

Comment:
I would argue, that this is not true, it is a rare scenario when a gem does not ship a source distribution.
We only found one during the development.

Reply:
> We cannot force Ruby platform too: this requires modifications to the lock file

We should not touch the lock file - ever - as that is the only source of truth for us, if we change it, it's going to be hard arguing about incorrect platform targets, bad resolution (see the issue I linked elsewhere) to the consumer base.

Reply:
What I potentially don't agree with is the warning statement as argued by @slimreaper35 for one particular reason - I don't see any further scanning on dupes in case `allow_binary == False`, IOW there could be an identical dependency defined with platform `ruby` withing the lockfile along with the platform specific ones, couldn't it (for the same reasons as Python's wheels - GPU bound builds, complex and long builds) ?
If I'm not mistaken in that reasoning, then by having at least the `ruby` one, a subsequent build would still pass and so we ought not scare the user with such statements; and if it so happens that they'd fail the build, then we can point them to the docs and say there's a flag for binary deps.

Reply:
No, it won't pass, at least not always. It started failing with nokogiri despite nokogiri having ruby version. If a version is present in a lockfile it seems to be needed by Bundler. If it were the case I would not have noticed this at all. I don't like the idea of silently falling back to ruby platform. Binary implementations exist for a reason, and usually this reason is performance. With silent ruby fallback we'll end up with users having noticeably worse performance by default. And since everything will work they will probably grumble, but accept it. And then some time in the future they will discover that someone made a suboptimal choice for them and will be very upset.

Reply:
> Binary implementations exist for a reason, and usually this reason is performance. With silent ruby fallback we'll end up with users having noticeably worse performance by default. And since everything will work they will probably grumble, but accept it. And then some time in the future they will discover that someone made a suboptimal choice for them and will be very upset.

Not a problem IMO as there's going to be a documented CLI flag to avert this situation.

> If a version is present in a lockfile it seems to be needed by Bundler

Fair enough, that makes sense. I guess I was more focused on the parity with our `pip` module so that we don't try downloading any binary packages (unless specified in the lockfile explicitly) without the CLI flag, but admittedly the situation with Bundler is different.

Reply:
> Not a problem IMO as there's going to be a documented CLI flag

A user could be someone other than a package builder. We have a pretty long supply chain here, there are many settings along it, I would not want to introduce another one with a surprising choice of defaults. I think this is something we can and should decide on on behalf of the users. This choice provides experience closer to one of installing dependencies on a connected machine, most users would expect that.

Another alternative I thought about is to build everything _including binary dependencies_ ourselves. This looks to be out of scope for cachi2: while we could download sources someone else would need to determine the right build sequence etc. This feels like a task for a proper package-manager. The next best thing is to try and replicate regular installation as close as possible. We can do this by downloading all binaries that Bundler deems necessary after a user agrees to that. Highlighting this is important because it is very easy to overlook a performance setting in the middle of a pipeline.